### PR TITLE
[WIP] POC: kubeadm: extract public API

### PR DESCRIPTION
**Pull request by rosti**
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:

This is a Proof-Of-Concept PR. It's not meant to merge (at least as is).
It was made to investigate and better imagine the extraction of the kubeadm public APIs into a more slim and easy to vendor way.

A few points to this:
- This is almost certainly not going to be the desired location of the APIs, but once this split is figured out, then it's trivial to move them.
- We'll probably do some discussing, update the kubeadm config KEP and then proceed with the final implementation (based on lessons learned here).
- Feedback (especially from vendoring users) is greatly appreciated.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/priority important-soon
/assign @neolit123
/assign @fabriziopandini
/assign @timothysc 

**Does this PR introduce a user-facing change?**:

```release-note
TODO
```

### CODE REVIEW

**Review by neolit123 — COMMENTED** [review]
i'm +1 on this move as is.

> Feedback (especially from vendoring users) is greatly appreciated.

it will break users who are importing k/k for the sake of having the public kubeadm types, but they should not be doing it anyhow. on the other hand, this change will help them as they will only get the public types, without the internal types.

also, we can move these public types to k/kubeadm/config today without the staging process (and sync manually).

**Comment by rosti** [comment]
> it will break users who are importing k/k for the sake of having the public kubeadm types, but they should not be doing it anyhow. on the other hand, this change will help them as they will only get the public types, without the internal types.

My idea is to have the new path as a drop-in replacement of the old import path. Hence, if you only use the public types (which is what one should do), a simple import path change should be straight forward enough to fix "broken" vendoring users.
 
> also, we can move these public types to k/kubeadm/config today without the staging process (and sync manually).

Certainly, they should not be at the spot where this PR puts them. I myself prefer the staging option for now (as opposed to having a manually synced copy in `k/kubeadm/config`). I want possible changes in both the internal and external types (such as a new config version) to go with a single PR in a single repo and staging would allow us to do precisely that.

Also, we need to make sure, that wherever we put them, we won't move them in the foreseeable future, so that we don't get annoying to vendoring users.
Finally, `k8s.io/kubeadm/api/config/v1beta2` looks like a good package name to me? WDYT?

**Comment by rosti** [comment]
Another thing that I am pondering over is what to do with the defaulting constants. We seem to have the following options:
- Leave them in `k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1betaX`.
- Move just the consts to the public APIs.
- Move just the consts to the kubeadm `constants` package and make that package more standalone (easy to vendor).
- Some mix of the above, which amounts to a reshuffling of the consts based on their specifics.

**Comment by neolit123** [comment]
> Also, we need to make sure, that wherever we put them, we won't move them in the foreseeable future, so that we don't get annoying to vendoring users.
Finally, k8s.io/kubeadm/api/config/v1beta2 looks like a good package name to me? WDYT?

there was a discussion about group name standardization in the #api-reviews channel on slack yesterday.
currently things are a bit chaotic.

for instance, kube-proxy and kubelet have "config" in their group, but not kubeadm.
`kubeproxy.config.k8s.io`
`kubeadm.k8s.io`

even if we lack "config" in the group, i think it make sense to put this under "api/config" and potentially have the machine-readable API under "api/output" with a group of `kubeadm.output.k8s.io`

**Comment by neolit123** [comment]
> Leave them in k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1betaX.
> Move just the consts to the public APIs.
> Move just the consts to the kubeadm constants package and make that package more standalone (easy to vendor).
> Some mix of the above, which amounts to a reshuffling of the consts based on their specifics.

i'm leaning towards having defaults.go under the public API.

having a look at:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/v1beta2/defaults.go

```
constants.DefaultControlPlaneTimeout
constants.DefaultTokenDuration
constants.DefaultTokenUsage
constants.DefaultTokenGroups
```

it imports "k8s.io/kubernetes/cmd/kubeadm/app/constants" only for a few constants that can live in the public type under "defaults.go". the rest of kubeadm can import the public API to obtain the constants as it already imports the latest public API. currently we are doing it the other way around.

also, having defaults.go in the public types will allow the users to default their structures using functions such as `SetDefaults_ClusterConfiguration`

**Comment by detiber** [comment]
> Certainly, they should not be at the spot where this PR puts them. I myself prefer the staging option for now (as opposed to having a manually synced copy in k/kubeadm/config). I want possible changes in both the internal and external types (such as a new config version) to go with a single PR in a single repo and staging would allow us to do precisely that.

+1 to staging, that would allow us to vendor the types without having to vendor k/k

I'm also a +1 on moving the defaults/constants to the public API

**Comment by yastij** [comment]
+1 on moving the consts and defaults, this would avoid depending on k/k or duplication for clients that want to use them

**Comment by neolit123** [comment]
> having defaults.go in the public types will allow the users to default their structures using functions such as SetDefaults_ClusterConfiguration.

which will surely not pass golint. queue the generator issues.

**Comment by yastij** [comment]
@neolit123 - we can still ignore lint errors, or fix generation long-term

**Comment by neolit123** [comment]
ignoring lint errors is annoying, while fixing the generators needs convincing api-machinery.

**Comment by rosti** [comment]
I am not quite sure why we want the defaulting code in there. The only viable case for this is for some tools to read existing kubeadm config, default it and try to predict what kubeadm would do. This, however, ignores dynamic defaulting and produces only "half-baked" config (since only static defaulting is applied).
In general, it is not needed to kubeadm config generators, since they would want to fill in only the fields, in which they are interested. Leaving everything else empty would cause kubeadm to default it (both statically and dynamically).

To me, validation and conversion code is of much higher value than defaulting. And that's the real loss for me.

In addition to that, we can always move the config default consts from the versioned APIs to the constants package and make that public and as standalone as possible (which it already is... to some extent).

**Comment by yastij** [comment]
> ignoring lint errors is annoying, while fixing the generators needs convincing api-machinery.

I agree that it's not ideal, but I wouldn't block on it. We can have something short-term (ignoring the lints) while working the long-term solution.

**Comment by fabriziopandini** [comment]
Thanks @rosti, great to see this moving so fast!

+1 on this move as is (without constants/defaults/conversions, see note below *)
+1 to staging, that would allow us to vendor the types without having to vendor k/k
+1 on k8s.io/kubeadm/api/config/v1beta2 as a package name
+1 on kubeadm.config.k8s.io as a group name

* few reasons make me think constant and defaulting are not :
- kubeadm config generators should not be required to set defaults values; fill in only the fields, in which they are interested and leave to kubeadm the responsibility to apply defaults.
- functions such as SetDefaults_ClusterConfiguration have little value without all the dynamic defaulting logic
- kubeadm config generators should not be required to do conversions; fill in one version, and let kubeadm do conversions if necessary.
- as of today also kubelet and kube-proxy are not including constants/defaults/conversions in their public API

**Comment by neolit123** [comment]
to clarify on "staging". we cannot use the official k8s staging process as it will stomp the existing contents of the k/kubeadm repo.

**Comment by detiber** [comment]
@neolit123, maybe a new repo: `k/kubeadm-api`?

**Comment by neolit123** [comment]
the consensus was that we can/should avoid the staging process.
the only benefits it has are auto-tags and release-branches.

i think ideally the API and kubeadm itself should live both in k/kubeadm and tagging and branches should be managed manually.

before kubeadm moves out, we can move the public types to k/kubeadm and introduce new API versions if new API versions are out with a k8s release.

i have no information of what the kubeproxy and kubelet long term plans are with respect to moving out the components them self, but i expect them to be in e.g. kubernetes/kube-proxy and not kubernetes/kube-proxy-component.

**Comment by timothysc** [comment]
Let's chat about this today, there are a bunch of logistical issues here.

**Comment by timothysc** [comment]
@rosti - could you summarize the decisions made today please.

**Comment by rosti** [comment]
Summerizing the discussion from [yesterday's kubeadm office hours]() as follows:

- We should start by moving the extracted APIs to k/kubeadm
  - Staging is a clear solution for now, but the publishing bot will stomp the existing contents of k/kubeadm and this is not an option for us ATM.
  - Another possibility is to use a separate k/kubeadm-api repository just for the APIs. This, however, makes the release process more complex and prone to errors. It also creates too much segmentation.
  - We have to be careful about possible policies regarding the change of location of released public beta APIs.
    - No code should be change, so if users are importing `k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2` and they use just the public APIs, then they should simply have to replace the imports with the new paths and not change any actual code.
    - Ask @liggitt for an opinion if we are breaking some policy or not.

- Defaulting needs further investigation, but so far the predominant idea is to keep defaulting code and consts out of the exported public APIs.
  - This is what is done by the rest of the exposed kubernetes APIs.
  - Need to verify that decision with more of the folks who are interested in vendoring the kubeadm APIs.
    - @chuckha , @detiber , @dlipovetsky , @BenTheElder

**Comment by detiber** [comment]
@rosti  +1 from me

**Comment by liggitt** [comment]
> We have to be careful about possible policies regarding the change of location of released public beta APIs.
> * No code should be changed, so if users are importing `k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2` and they use just the public APIs, then they should simply have to replace the imports with the new paths and not change any actual code.
> * Ask @liggitt for an opinion if we are breaking some policy or not.

The published policies are concerned with serialized content and behavior, not location of go types.

**Comment by rosti** [comment]
This POC has served its purpose. Closing it.
